python/lsst/daf/butler/core/serverModels.py

```diff
@@ -0,0 +1,286 @@
+Collections = Field(
+    None,
+    title="An expression that identifies the collections to search.",
+)
+Datasets = Field(
+    None,
+    title="An expression that identifies dataset types to search (must not match all datasets).",
+)
+OptionalDimensions = Field(
+    None,
+    title="Relevant dimensions to include.",
+    example=["detector", "physical_filter"],
+)
+Dimensions = Field(
+    ...,
+    title="Relevant dimensions to include.",
+    example=["detector", "physical_filter"],
+)
+DataId = Field(
+    None,
+    title="Data ID to constrain the query.",
+)
+FindFirst = Field(
+    False,
+    title="Control whether only first matching dataset ref or type is returned.",
+)
+Components = Field(
+    None,
+    title="Control how expressions apply to components.",
+)
+Bind = Field(
+    None,
+    title="Mapping to use to inject values into the WHERE parameter clause.",
+)
+Check = Field(
+    True,
+    title="Control whether to check the query for consistency.",
+)
+KeywordArgs = Field(
+    None,
+    title="Additional parameters to use when standardizing the supplied data ID.",
+)
+
+
+class QueryBaseModel(BaseModel):
+    """Base model for all query models."""
+
+    @validator("keyword_args", check_fields=False)
+    def check_keyword_args(cls, v, values) -> Optional[SimpleDataId]:  # type: ignore # noqa: N805
+        """Convert kwargs into None if empty.
+
+        This retains the property at its default value and can therefore
+        remove it from serialization.
+
+        The validator will be ignored if the subclass does not have this
+        property in its model.
+        """
+        if not v:
+            return None
+        return v
+
+    def kwargs(self) -> SimpleDataId:
+        """Return keyword args, converting None to a `dict`.
+
+        Returns
+        -------
+        kwargs : `dict`
+            The keword arguments stored in the model. `None` is converted
+            to an empty dict. Returns empty dict if the ``keyword_args``
+            property is not defined.
+        """
+        try:
+            # mypy does not know about the except
+            kwargs = self.keyword_args  # type: ignore
+        except AttributeError:
+            kwargs = {}
+        if kwargs is None:
+            return {}
+        return kwargs
+
+
+class QueryDatasetsModel(QueryBaseModel):
+    """Information needed for a registry dataset query."""
+
+    datasetType: ExpressionQueryParameter = Field(
+        ...,
+        title="Dataset types to query. Can match all."
+    )
+    collections: Optional[ExpressionQueryParameter] = Collections
+    dimensions: Optional[List[str]] = OptionalDimensions
+    dataId: Optional[SerializedDataCoordinate] = DataId
+    where: Optional[str] = Where
+    findFirst: bool = FindFirst
+    components: Optional[bool] = Components
+    bind: Optional[BindType] = Bind
+    check: bool = Check
+    keyword_args: Optional[SimpleDataId] = KeywordArgs  # mypy refuses to allow kwargs in model
```

Comment:
Can't you lift this up to the parent class since it's common in all subclasses? Or is that what the comment is about?

Reply:
I was going to build all the models and then work out which bits needed to be in the base.